From cef4664e7796afc169aef2d9c58c8a1517bf54d9 Mon Sep 17 00:00:00 2001
From: Anthony Koo <Anthony.Koo@amd.com>
Date: Tue, 31 May 2016 12:04:02 -0400
Subject: [PATCH 0514/1722] drm/amd/dal: VSR Downscale

Implement DC code to have downscale ratio limit as
an ASIC capability

Signed-off-by: Anthony Koo <Anthony.Koo@amd.com>
Acked-by: Jordan Lazare <Jordan.Lazare@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 drivers/gpu/drm/amd/dal/dc/adapter/adapter_service.c                | 6 ++++++
 drivers/gpu/drm/amd/dal/dc/asic_capability/asic_capability.c        | 1 +
 .../gpu/drm/amd/dal/dc/asic_capability/carrizo_asic_capability.c    | 1 +
 drivers/gpu/drm/amd/dal/dc/core/dc.c                                | 3 +++
 drivers/gpu/drm/amd/dal/dc/dc.h                                     | 1 +
 drivers/gpu/drm/amd/dal/include/adapter_service_interface.h         | 3 +++
 drivers/gpu/drm/amd/dal/include/asic_capability_types.h             | 1 +
 7 files changed, 16 insertions(+)

diff --git a/drivers/gpu/drm/amd/dal/dc/adapter/adapter_service.c b/drivers/gpu/drm/amd/dal/dc/adapter/adapter_service.c
index ffc9299..760767e 100644
--- a/drivers/gpu/drm/amd/dal/dc/adapter/adapter_service.c
+++ b/drivers/gpu/drm/amd/dal/dc/adapter/adapter_service.c
@@ -2049,6 +2049,12 @@ uint32_t dal_adapter_service_get_num_of_underlays(
 	return as->asic_cap->data[ASIC_DATA_NUM_OF_VIDEO_PLANES];
 }
 
+uint32_t dal_adapter_service_get_downscale_limit(
+		struct adapter_service *as)
+{
+	return as->asic_cap->data[ASIC_DATA_DOWNSCALE_LIMIT];
+}
+
 bool dal_adapter_service_get_encoder_cap_info(
 		struct adapter_service *as,
 		struct graphics_object_id id,
diff --git a/drivers/gpu/drm/amd/dal/dc/asic_capability/asic_capability.c b/drivers/gpu/drm/amd/dal/dc/asic_capability/asic_capability.c
index aeabfc6..3456ca04 100644
--- a/drivers/gpu/drm/amd/dal/dc/asic_capability/asic_capability.c
+++ b/drivers/gpu/drm/amd/dal/dc/asic_capability/asic_capability.c
@@ -72,6 +72,7 @@ static bool construct(
 	cap->data[ASIC_DATA_SUPPORTED_HDMI_CONNECTION_NUM] = 1;
 	cap->data[ASIC_DATA_NUM_OF_VIDEO_PLANES] = 0;
 	cap->data[ASIC_DATA_DEFAULT_I2C_SPEED_IN_KHZ] = 25;
+	cap->data[ASIC_DATA_DOWNSCALE_LIMIT] = 200;
 
 	/* ASIC basic capability */
 	cap->caps.UNDERSCAN_FOR_HDMI_ONLY = true;
diff --git a/drivers/gpu/drm/amd/dal/dc/asic_capability/carrizo_asic_capability.c b/drivers/gpu/drm/amd/dal/dc/asic_capability/carrizo_asic_capability.c
index 4f2e6b8..340c1f1 100644
--- a/drivers/gpu/drm/amd/dal/dc/asic_capability/carrizo_asic_capability.c
+++ b/drivers/gpu/drm/amd/dal/dc/asic_capability/carrizo_asic_capability.c
@@ -65,6 +65,7 @@ void carrizo_asic_capability_create(struct asic_capability *cap,
 	cap->data[ASIC_DATA_DEFAULT_I2C_SPEED_IN_KHZ] = 100;
 	cap->data[ASIC_DATA_NUM_OF_VIDEO_PLANES] = 1;
 	cap->data[ASIC_DATA_SUPPORTED_HDMI_CONNECTION_NUM] = 3;
+	cap->data[ASIC_DATA_DOWNSCALE_LIMIT] = 150;
 
 	/* ASIC basic capability */
 	cap->caps.IS_FUSION = true;
diff --git a/drivers/gpu/drm/amd/dal/dc/core/dc.c b/drivers/gpu/drm/amd/dal/dc/core/dc.c
index 0ca1d5a..e7f37bc 100644
--- a/drivers/gpu/drm/amd/dal/dc/core/dc.c
+++ b/drivers/gpu/drm/amd/dal/dc/core/dc.c
@@ -433,6 +433,9 @@ struct dc *dc_create(const struct dc_init_data *init_params)
 			core_dc->res_pool.stream_enc_count);
 	core_dc->public.caps.max_links = core_dc->link_count;
 	core_dc->public.caps.max_audios = core_dc->res_pool.audio_count;
+        core_dc->public.caps.max_downscale_ratio =
+                        dal_adapter_service_get_downscale_limit(
+                                        core_dc->res_pool->adapter_srv);
 
 	dal_logger_write(core_dc->ctx->logger,
 			LOG_MAJOR_INTERFACE_TRACE,
diff --git a/drivers/gpu/drm/amd/dal/dc/dc.h b/drivers/gpu/drm/amd/dal/dc/dc.h
index 804a5a9..61d966d 100644
--- a/drivers/gpu/drm/amd/dal/dc/dc.h
+++ b/drivers/gpu/drm/amd/dal/dc/dc.h
@@ -46,6 +46,7 @@ struct dc_caps {
 	uint32_t max_links;
 	uint32_t max_audios;
 	uint32_t max_underlays;
+	uint32_t max_downscale_ratio;
 };
 
 struct dc;
diff --git a/drivers/gpu/drm/amd/dal/include/adapter_service_interface.h b/drivers/gpu/drm/amd/dal/include/adapter_service_interface.h
index 59f395b..87c834a 100644
--- a/drivers/gpu/drm/amd/dal/include/adapter_service_interface.h
+++ b/drivers/gpu/drm/amd/dal/include/adapter_service_interface.h
@@ -616,6 +616,9 @@ uint32_t dal_adapter_service_get_num_of_path_per_dp_mst_connector(
 uint32_t dal_adapter_service_get_num_of_underlays(
 		struct adapter_service *as);
 
+uint32_t dal_adapter_service_get_downscale_limit(
+		struct adapter_service *as);
+
 bool dal_adapter_service_get_encoder_cap_info(
 		struct adapter_service *as,
 		struct graphics_object_id id,
diff --git a/drivers/gpu/drm/amd/dal/include/asic_capability_types.h b/drivers/gpu/drm/amd/dal/include/asic_capability_types.h
index 56fdcd8..7841662 100644
--- a/drivers/gpu/drm/amd/dal/include/asic_capability_types.h
+++ b/drivers/gpu/drm/amd/dal/include/asic_capability_types.h
@@ -109,6 +109,7 @@ enum asic_data {
 	ASIC_DATA_MIN_DISPCLK_FOR_UNDERSCAN,
 	ASIC_DATA_NUM_OF_VIDEO_PLANES,
 	ASIC_DATA_DEFAULT_I2C_SPEED_IN_KHZ,
+	ASIC_DATA_DOWNSCALE_LIMIT,
 	ASIC_DATA_MAX_NUMBER /* end of enum */
 };
 
-- 
2.7.4

